<a href="https://colab.research.google.com/github/khs1414/Kaustubh21CSU043-RL/blob/main/rl_22_08_2023.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Q1. Student grades and absence
Download the dataset from Kaggle:

https://www.kaggle.com/datasets/uciml/student-alcohol-consumption

Compute the probability that a student obtains 80% or more in mathematics, given he/she misses atleast 10 classes. Do this in Python.

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('student-mat.csv')
df.head(3)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home  teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home    other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home    other  ...   

  famrel freetime  goout  Dalc  Walc health absences G1 G2  G3  
0      4        3      4     1     1      3        6  5  6   6  
1      5        3      3     1     1      3        4  5  5   6  
2      4        3      2     2     3      3       10  7  8  10  

[3 rows x 33 columns]

Add a boolean column called grade_A noting if a student achieved 80% or higher as a final score. Original values are on a 0–20 scale so we multiply by 5.

In [4]:
df.head(2)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home  teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home    other  ...   

  famrel freetime  goout  Dalc  Walc health absences G1 G2 G3  
0      4        3      4     1     1      3        6  5  6  6  
1      5        3      3     1     1      3        4  5  5  6  

[2 rows x 33 columns]

In [5]:
#Add a boolean column called grade_A noting
#if a student achieved 80% or higher as a final score.
#Original values are on a 0–20 scale so we multiply by 5.
df['grade_A'] = np.where(df['G3']*5 >= 80, 1, 0)

Make another boolean column called high_absenses with a value of 1 if a student missed 10 or more classes.

In [6]:
df['high_absenses'] = np.where(df['absences'] >= 10, 1, 0)

In [7]:
df.head(2)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob     Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home  teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home    other  ...   

  goout Dalc  Walc  health  absences G1 G2 G3 grade_A high_absenses  
0     4    1     1       3         6  5  6  6       0             0  
1     3    1     1       3         4  5  5  6       0             0  

[2 rows x 35 columns]

Add one more column to make building a pivot table easier.

In [8]:
df['count'] = 1

In [9]:
df = df[['grade_A','high_absenses','count']]
df.head()


grade_A  high_absenses  count
0        0              0      1
1        0              0      1
2        0              1      1
3        0              0      1
4        0              0      1

In [10]:
pd.pivot_table(
    df,
    values='count',
    index=['grade_A'],
    columns=['high_absenses'],
    aggfunc=np.size,
    fill_value=0
)

high_absenses    0   1
grade_A               
0              277  78
1               35   5

In our case:
P(A) is the probability of a grade of 80% or greater.
P(B) is the probability of missing 10 or more classes.
P(A|B) is the probability of a 80%+ grade, given missing 10 or more classes.

Calculations of parts:
P(A) = (35 + 5) / (35 + 5 + 277 + 78) = 0.10126582278481013
P(B) = (78 + 5) / (35 + 5 + 277 + 78) = 0.21012658227848102
P(A ∩ B) = 5 / (35 + 5 + 277 + 78) = 0.012658227848101266

P(A|B) = 0.012658227848101266/ 0.21012658227848102= 0.06

There we have it. The probability of getting at least an 80% final grade, given missing 10 or more classes is 6%.

Ques 2. **THE BIRTHDAY PROBLEM**
Compute the probability of getting a minimum of one overlapping birthday in a random group of 23 peoples.
Because ontaining random sample again and again is tedious task, we can do simulations on a computer with assumptions.
The birthdays are independent of each other. Each possible birthday has the same probability. There are only 365 possible birthdays (not 366, as ignoring the leap year.
Hint : in other words, we're modelling the process as drawing 23 independent samples from a discrete uniform distribution with parameter n = 365.

In [11]:
from random import randint


NUM_PEOPLE = 23
NUM_POSSIBLE_BIRTHDAYS = 365
NUM_TRIALS = 10000


def generate_random_birthday():
    birthday = randint(1, NUM_POSSIBLE_BIRTHDAYS)
    return birthday


def generate_k_birthdays(k):
    birthdays = [generate_random_birthday() for _ in range(k)]
    return birthdays


def aloc(birthdays):
    unique_birthdays = set(birthdays)

    num_birthdays = len(birthdays)
    num_unique_birthdays = len(unique_birthdays)
    has_coincidence = (num_birthdays != num_unique_birthdays)

    return has_coincidence


def estimate_p_aloc():
    num_aloc = 0
    for _ in range(NUM_TRIALS):
        birthdays = generate_k_birthdays(NUM_PEOPLE)
        has_coincidence = aloc(birthdays)
        if has_coincidence:
            num_aloc += 1

    p_aloc = num_aloc / NUM_TRIALS
    return p_aloc


p_aloc = estimate_p_aloc()
print(f"Estimated P(ALOC) after {NUM_TRIALS} trials: {p_aloc}")

Estimated P(ALOC) after 10000 trials: 0.5016


**Ques 3. The weight of certain species of frog is uniformly distributed from 15 and 25 grams. if you randomly select a frog, what is the probability that the frog weights between 17 and 19 grams.**

In [12]:
from scipy.stats import uniform
uniform.cdf(x=19, loc = 15, scale =10) -uniform.cdf(x=17, loc =15, scale=10)

0.2

**Ques 4. Telecommunication Industry**
According to the Telecommunication Industry the average monthly cell phone bill is Rs. 1000 with a standard deviation of Rs. 200.

What is the probability that a randomly selected cell phone bill is more than Rs 1200? What is the probability that a randomly selected cell phone bill is between Rs 750 and Rs 1200? What is the probability that a randomly selected cell phone bill is no more than Rs 650? What is the amount above which lies top 15% of cell phone bills? What is the amount below which lies bottom 25% of cell phone bills?

Note: This is a problem of normal probability distribution. Though the distribution is not mentioned, in absence of any other information we assume normality in the population.

In [15]:
from scipy import stats
# Given information
mean = 1000
std_dev = 200

# Probability that a randomly selected cell phone bill is more than Rs 1200
prob_above_1200 = 1 - stats.norm.cdf(1200, mean, std_dev)
print("Probability of a bill being more than Rs 1200:", prob_above_1200)

# Probability that a randomly selected cell phone bill is between Rs 750 and Rs 1200
prob_between_750_1200 = stats.norm.cdf(1200, mean, std_dev) - stats.norm.cdf(750, mean, std_dev)
print("Probability of a bill being between Rs 750 and Rs 1200:", prob_between_750_1200)

# Probability that a randomly selected cell phone bill is no more than Rs 650
prob_no_more_than_650 = stats.norm.cdf(650, mean, std_dev)
print("Probability of a bill being no more than Rs 650:", prob_no_more_than_650)

# Find the amount above which lies the top 15% of cell phone bills
amount_above_top_15 = stats.norm.ppf(0.85, mean, std_dev)
print("Amount above which lies the top 15% of bills:", amount_above_top_15)

# Find the amount below which lies the bottom 25% of cell phone bills
amount_below_bottom_25 = stats.norm.ppf(0.25, mean, std_dev)
print("Amount below which lies the bottom 25% of bills:", amount_below_bottom_25)


Probability of a bill being more than Rs 1200: 0.15865525393145707
Probability of a bill being between Rs 750 and Rs 1200: 0.7356949724016876
Probability of a bill being no more than Rs 650: 0.040059156863817086
Amount above which lies the top 15% of bills: 1207.286677898758
Amount below which lies the bottom 25% of bills: 865.1020499607837


Q5. Fruit problem
Suppose we own a fruit shop and on an average 3 customers arrive in the shop every 10 minutes. The mean rate here is 3 or λ = 3. Poisson probability distributions can help us answer questions like what is the probability that 5 customers will arrive in the next 10 mins?

In [16]:
from scipy.stats import poisson
poisson.cdf(k=5, mu=3)

0.9160820579686966